File: Tutorial1-2DaggerBasics2/src/main/java/com/example/tutorial1basics3/di/AppContextModule.java
package com.example.tutorial1basics3.di;

import com.example.tutorial1basics3.MyApplication;
import com.example.tutorial1basics3.model.SensorController;
import com.example.tutorial1basics3.model.User;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public class AppContextModule {
    private final MyApplication mMyApplication;

    public AppContextModule(MyApplication context) {
        mMyApplication = context;
    }

    // @Singleton annotation for application object is not necessary since it's already a singleton object of Android
    @Singleton
    @Provides
    MyApplication provideMyApplication() {
        return mMyApplication;
    }

    // This is a singleton object at application level
    @Singleton
    @Provides
    User provideUser() {
        return new User();
    }

    // This is not a singleton object
    @Provides
    SensorController provideSensorController() {
        return new SensorController(mMyApplication);
    }


}

File: Tutorial1-2DaggerBasics2/src/main/java/com/example/tutorial1basics3/di/AppContextModule.kt
package com.example.tutorial1basics3.di

import com.example.tutorial1basics3.MyApplication
import com.example.tutorial1basics3.model.SensorController
import com.example.tutorial1basics3.model.User
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
class AppContextModule(private val myApplication: MyApplication) {

    // @Singleton annotation for application object is not necessary since it's already a singleton object of Android
    @Singleton
    @Provides
    fun provideMyApplication(): MyApplication {
        return myApplication
    }

    // This is a singleton object at application level
    @Singleton
    @Provides
    fun provideUser(): User {
        return User()
    }

    // This is NOT a singleton object
    @Provides
    fun provideSensorController(): SensorController {
        return SensorController(myApplication)
    }

}